packages/astro/src/vite-plugin-build-css/index.ts

```diff
@@ -38,193 +15,142 @@ interface PluginOptions {
+		name: '@astrojs/rollup-plugin-build-css',
 
 		configResolved(resolvedConfig) {
-			// Our plugin needs to run before `vite:css-post` which does a lot of what we do
-			// for bundling CSS, but since we need to control CSS we should go first.
-			// We move to right before the vite:css-post plugin so that things like the
-			// Vue plugin go before us.
+			// Our plugin needs to run before `vite:css-post` because we have to modify
+			// The bundles before vite:css-post sees them. We can remove this code
+			// after this bug is fixed: https://github.com/vitejs/vite/issues/8330
 			const plugins = resolvedConfig.plugins as VitePlugin[];
 			const viteCSSPostIndex = resolvedConfig.plugins.findIndex((p) => p.name === 'vite:css-post');
 			if (viteCSSPostIndex !== -1) {
-				const viteCSSPost = plugins[viteCSSPostIndex];
-				// Prevent this plugin's bundling behavior from running since we need to
-				// do that ourselves in order to handle updating the HTML.
-				delete viteCSSPost.renderChunk;
-				delete viteCSSPost.generateBundle;
-
 				// Move our plugin to be right before this one.
-				const ourIndex = plugins.findIndex((p) => p.name === PLUGIN_NAME);
+				const ourIndex = plugins.findIndex((p) => p.name === '@astrojs/rollup-plugin-build-css');
 				const ourPlugin = plugins[ourIndex];
 
 				// Remove us from where we are now and place us right before the viteCSSPost plugin
 				plugins.splice(ourIndex, 1);
 				plugins.splice(viteCSSPostIndex - 1, 0, ourPlugin);
 			}
 		},
-		async resolveId(id) {
-			if (isPageStyleVirtualModule(id)) {
-				return id;
-			}
-			if (isStyleVirtualModule(id)) {
-				return id;
-			}
-			return undefined;
-		},
-
-		async transform(value, id) {
-			if (isStyleVirtualModule(id)) {
-				styleSourceMap.set(id, value);
-			}
-			if (isCSSRequest(id)) {
-				styleSourceMap.set(id, value);
-			}
-			return null;
-		},
-
-		async renderChunk(_code, chunk) {
-			if (options.target === 'server') return null;
 
-			let chunkCSS = '';
-			let isPureCSS = true;
-			for (const [id] of Object.entries(chunk.modules)) {
-				if (!isCSSRequest(id) && !isPageStyleVirtualModule(id)) {
-					isPureCSS = false;
-				}
-				if (styleSourceMap.has(id)) {
-					chunkCSS += styleSourceMap.get(id)!;
+		outputOptions(outputOptions) {
+			const manualChunks = outputOptions.manualChunks || Function.prototype;
+			outputOptions.manualChunks = function(id, ...args) {
+				// Defer to user-provided `manualChunks`, if it was provided.
+				if(typeof manualChunks == 'object') {
+					if(id in manualChunks) {
+						return manualChunks[id];
+					}
+				} else if(typeof manualChunks === 'function') {
+					const outid = manualChunks.call(this, id, ...args);
+					if(outid) {
+						return outid;
+					}
 				}
-			}
 
-			if (!chunkCSS) return null; // don’t output empty .css files
+				// For CSS, create a hash of all of the pages that use it.
+				// This causes CSS to be built into shared chunks when used by multiple pages.
```

Comment:
Nice!!!  This will be a huge help when it comes time to revisit a SPA mode

Reply:
YESSSS. Very glad to see us going back to how we did this pre-Vite.